### ISSUE TITLE
Opinion: NVDA Should Provide Its Own Image Captcha Solving Service

### ISSUE THREAD

**Issue by Elshara** [nvaccess/nvda#5979]
Too many web pages online still use captcha. For those who don't know what captcha is, it is a way of inaccessible image confirmation to try to prove humans are not web bots. It is antiproductive to all blind people with vision impairment and as such, prevents people using NVDA to be able to accessibly fill out web forms. Web forms that are required to do something basic like set up an email account, fill out an online order, sign up for a web site or news service and be able to make use of online website submission tools to directories and search engines if you run your own online business.
Some image tools, obviously cannot be accessible because NVDA does not know how to navigate slider based content at all. This means if NVDA cannot recognize shapes, it is therefore unable to complete a captcha. This should be changed given NVDA's OCR functionality to provide text captions for images. If this is implemented, NVDA may be able to recognize flash based file uploaders as well.

**Comment by josephsl** [comment]
Hi, This is just me, but I’d rather let NVDA be an expert on what it does: screen reading, not an all-in-one access solution for computing needs. There is Webvisum and other such services that are known to work well with NVDA. Thanks.

**Comment by derekriemer** [comment]
Note that providing OCR capabilities for solving captcha's is 
impossible. The reasoning behind captcha is to create something a 
computer can't solve. If OCR could read the captcha, it would defeat the 
purpose


On 5/20/2016 8:23 AM, josephsl wrote:
> Hi, This is just me, but I’d rather let NVDA be an expert on what it 
> does: screen reading, not an all-in-one access solution for computing 
> needs. There is Webvisum and other such services that are known to 
> work well with NVDA. Thanks.
>
>
>
> From: Elshara Silverheart [mailto:notifications@github.com]
> Sent: Friday, May 20, 2016 3:29 AM
> To: nvaccess/nvda
> Subject: [nvaccess/nvda] Opinion: NVDA Should Provide Its Own Image 
> Captcha Solving Service (#5979)
>
>
>
> Too many web pages online still use captcha. For those who don't know 
> what captcha is, it is a way of inaccessible image confirmation to try 
> to prove humans are not web bots. It is antiproductive to all blind 
> people with vision impairment and as such, prevents people using NVDA 
> to be able to accessibly fill out web forms. Web forms that are 
> required to do something basic like set up an email account, fill out 
> an online order, sign up for a web site or news service and be able to 
> make use of online website submission tools to directories and search 
> engines if you run your own online business.
> Some image tools, obviously cannot be accessible because NVDA does not 
> know how to navigate slider based content at all. This means if NVDA 
> cannot recognize shapes, it is therefore unable to complete a captcha. 
> This should be changed given NVDA's OCR functionality to provide text 
> captions for images. If this is implemented, NVDA may be able to 
> recognize flash based file uploaders as well.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/5979> Image removed by sender.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/5979#issuecomment-220619637>
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

**Comment by Elshara** [comment]
1. Webvism is discontinued.
2. Captcha services cannot be decoded even if NVDA wasn't installed if
a blind person were to gain access to a product they are more than
able to work with if they had a working knowledge of how to get past
things captcha services assume someone is able to do without fail.
This means, it is most likely the fault of a screen reader to the
untrained eye if by chance, no audio is provided, and is a part of
making the web accessible. Especially when it comes to email account
creation, for people who don't have access to a phone, mobile or
accessibility wise. The job of a screen reader is to bring blind
people their independence when working on the internet, and this is
essential if we are to keep up with modern technology because captcha
is not going away any time soon. So in a sense, you are completely
missing the point on what OCR plus captcha recognition would be able
to do specifically and exclusively for the purposes of a browser add
on. All it would take would be for NVDA to recognize a captcha element
exists on a page by providing automatic OCR functionality into the
reading document so that not just captchas can be translated, but also
other images obtained through pdf documents, regardless of how the
text font is seen. If a captcha element is found, NVDA could report
whether or not it is accessible by clicking a check box or if it
requires a custom command to send the data through. Similar to how
webvisum worked, but different in a sense it would be built in even as
an add on to NVDA so people could make good use out of websites that
are made for humans, without the websites having to drastically change
captcha technology for screen reader accessibility. We have to deal
with all different types of pages on the web. Pages with no links, no
lists, no headings where they would be most useful. We shouldn't have
to rely on inaccessibility to get around something we love, to empower
our freedom online. That is exactly what captcha does, by telling us
we can't have something we know we could use otherwise.

**Comment by josephsl** [comment]
Hi,

1. Webvisum is back: http://www.webvisum.com/
2. If I understand you right, you are saying that screen readers should be responsible for web accessibility, including that of CAPTCHA. I'd like to kindly disagree: web accessibility is a collaborative work that involves constant communication and feedback between webmasters, users, browser vendors, screen reader developers, web standards committees and others. Also, there are audio CAPTCHAs that can be solved by blind folks (Google implemented  this and a number of other companies have as well). Another alternative is to provide a "fake" form field that humans would not see but bots would be fooled into filling it out, or provide a math question or other questions that can be solved by humans (not just images and audio, but other textual question(s)).

In summary, I guess this ticket came up possibly due to an impression that screen readers are an essential tool for accessing the internet. To some, it is true, but when you dive deeper into discussion of accessibility, you'll find that it is really a collaborative communication that drives web accessibility, not Python/C++ code. Thus I'd like to kindly request that we either turn this request down or defer it to another time (I'll not close this, as I think Mick and/or Jamie would like to offer their opinions and then close this one).

Thanks.

**Comment by derekriemer** [comment]
I would also like to explain Captcha in more detail.


1. Captcha is a computer automated test designed to make sure humans and 
only humans are using a service.

2. Captcha cannot *By Definition* be solved by a screen reader using OCR 
like we have in the OCR add-on. It is designed so computers cannot solve it.

3. Captcha solving services like in web Visum often send captcha's off 
to crowd computing services, like mechanical turk, where a human is paid 
some small amount like 3 cents per transaction to solve captchas, or do 
other tasks a computer cannot do.

Therefore, in order for NVDA to do this, it would require a constant 
source of extra revenue from users.
Thanks.
On 5/21/2016 2:56 PM, josephsl wrote:
>
> Hi,
>
>  1. Webvisum is back: http://www.webvisum.com/
>  2. If I understand you right, you are saying that screen readers
>     should be responsible for web accessibility, including that of
>     CAPTCHA. I'd like to kindly disagree: web accessibility is a
>     collaborative work that involves constant communication and
>     feedback between webmasters, users, browser vendors, screen reader
>     developers, web standards committees and others. Also, there are
>     audio CAPTCHAs that can be solved by blind folks (Google
>     implemented this and a number of other companies have as well).
>     Another alternative is to provide a "fake" form field that humans
>     would not see but bots would be fooled into filling it out, or
>     provide a math question or other questions that can be solved by
>     humans (not just images and audio, but other textual question(s)).
>
> In summary, I guess this ticket came up possibly due to an impression 
> that screen readers are an essential tool for accessing the internet. 
> To some, it is true, but when you dive deeper into discussion of 
> accessibility, you'll find that it is really a collaborative 
> communication that drives web accessibility, not Python/C++ code. Thus 
> I'd like to kindly request that we either turn this request down or 
> defer it to another time (I'll not close this, as I think Mick and/or 
> Jamie would like to offer their opinions and then close this one).
>
> Thanks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/5979#issuecomment-220799847>
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

**Comment by Elshara** [comment]
I understand a bit better now. Would you agree then, that to the
average NVDA user, services like Webvisum are widely known? Would you
also agree in theory, that if NVDA were to link to a website
supporting help with captchas, that it would be better for users to
use such a service who needs it? I do know the major providers does
offer an audio service, and that such an audio service is best used in
situations where no other service is available. Have you seen the
latest in captcha technology? Where humans have to slide images around
to complete a shape test? Or where there is a custom built third party
captcha that doesn't use a third party? Is it all related to how
developers build their code and nothing could be improved on our side
of the fence to work with the code built? Not every service is your
run of the mill common service out there on the internet. I'm just
trying to figure out ways where such services that don't offer audio,
and there are a lot of them out there, that I've personally had to use
a service such as webvisum for to get past, before firefox stopped
supporting the add on to run on any page at all, a die hard user of
it, to get past. My point being, the job of a screen reader is to
translate text to speech and this is no different.

**Comment by Brian1Gaff** [comment]
Or maybe all those using such outdated tech should get their act together 
and remove them...:-)
 Can you imagine what doing the function you say would do? It would in fact 
render them pointless to use, which might be good for us, but the creators 
might like to sue nvaccess for breaking their code and letting all the 
spammers in by the back door.
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal email to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
----- Original Message -----

**Comment by jcsteh** [comment]
Having in-built CAPTCHA solving would IMO definitly be a nice thing for a screen reader to have and is definitely in scope. However, creating such a solution is a massive undertaking and is not something we have the expertise and/or resources to create at this stage. If anyone is able to work on or provide such a solution, we'd love to collaborate. Until then, there's nothing further to discuss.

Closing as cantfix for now because it's unlikely to happen in the near future, but happy to reopen if this changes.

**Comment by Elshara** [comment]
Hey I'm keeping my eyes open on this one! Perhaps you should contact
web visum say you want to collaborate.
